src/backend/replication/walproposer.c

```diff
@@ -644,255 +720,615 @@ WalProposerRecovery(int leader, TimeLineID timeline, XLogRecPtr startpos, XLogRe
-						{
-							elog(WARNING, "WalKeeper has incompatible protocol version %d vs. %d",
-								wk->info.server.protocolVersion, SK_PROTOCOL_VERSION);
-							ResetConnection(i);
-						}
-						else
+/* Performs the logic for advancing the state machine of the 'i'th walkeeper, given that a certain
+ * set of events has occured. */
+static void
+AdvancePollState(int i, uint32 events)
+{
+	WalKeeper* wk = &walkeeper[i];
+
+	/* Continue polling all the while we don't need to wait.
+	 *
+	 * At the bottom of this function is "while (walkeeper[i].sockWaitState == WANTS_NO_WAIT)" */
+	do {
+		/* Helper variable for SPOLL_CONNECT/SPOLL_FIRST_CONNECT */
+		pgsocket previous_conn_sock = PGINVALID_SOCKET;
+
+		/* First, check that the events match what we're expecting. */
+		if (!(WaitKindAsEvents(wk->sockWaitState) & events))
+		{
+			elog(FATAL,
+				 "unexpected event for WalKeeper poll. Expected %s, found code %s (see: FormatEvents).",
+				 FormatWKSockWaitKind(wk->sockWaitState), FormatEvents(events));
+		}
+
+		/* Now that we've checked the event is ok, we'll actually run the thing we're looking for */
+		switch (wk->pollState)
+		{
+			/* If the polling corresponds to a "full" operation, we'll skip straight to that - we
+			 * don't actually need to poll here. */
+			case SPOLL_NONE:
+			case SPOLL_RETRY:
+				/* Equivalent to 'break', but more descriptive. */
+				goto ExecuteNextProtocolState;
+
+			/* Per the docs, this polling kind means that the state isn't expecting to be advanced
+			 * by calls to AdvancePollState. */
+			case SPOLL_IDLE:
+				elog(FATAL, "Attempted to advance idle polling walkeeper %s:%s in %s state",
+					 wk->host, wk->port, FormatWalKeeperState(wk->state));
+				break; /* actually unreachable, but prevents -Wimplicit-fallthrough */
+
+			/* Call PQconnectPoll to finalize the connection */
+			case SPOLL_CONNECT:
+				/* For everything but the first connection, we might need to reset the current
+				 * connection socket. This is because the connection's socket is allowed to change
+				 * across calls to PQconnectPoll.
+				 *
+				 * The tools we have to remove an event are a little slow, so it's better for us to
+				 * wait until after polling to see if we'll need to change the socket. */
+				previous_conn_sock = PQsocket(wk->conn);
+
+				/* And then fallthrough to the rest of the PQconnectPoll logic */
+				/* FALLTHRU */ // <- make gcc ok with this
+			case SPOLL_FIRST_CONNECT: ;
+				PostgresPollingStatusType result = PQconnectPoll(wk->conn);
+				pgsocket                  new_sock = PQsocket(wk->conn);
+
+				switch (result)
+				{
+					case PGRES_POLLING_OK:
+						elog(LOG, "Connected with node %s:%s", wk->host, wk->port);
+
+						/* If we're fully connected, we're good! We can move on to the next state */
+						wk->state = SS_HANDSHAKE_SEND;
+
+						/* If the socket changed while polling, we have to update the event */
+						if (new_sock != previous_conn_sock)
```

Comment:
That's a good question -- I don't know. I think it largely depends on what the event implementation does when a socket closes. If it disables all current events for that socket, then this might cause problems.

I'd originally written that because recreating the entire event set on each call to `ConnectPoll` seemed a little extreme to me, but maybe it's ok.